## Homework 18: Neural Network
**Conditions**:
- Dataset: fashion_mnist
- Metric: accuracy
1. Train simple neural network and get accuracy
2. Train models using homework conditions and compare results with simple neural network's 

**Goal**: make experiment on tuning different parameters of neural network

**Acceptance criteria**:  NONE

In [1]:
# Importing necessary libraries
import tensorflow as tf
from tensorflow import keras
import matplotlib.pyplot as plt
from keras.regularizers import l1, l2

2023-09-25 15:48:57.016278: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


## Creating simple neural network to obtain results

In [2]:
# Loading data
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

In [3]:
# Normalizing data
train_images = train_images / 255.0
test_images = test_images / 255.0

In [15]:
# Creating a model
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

In [16]:
# Compilingt the model
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# Training the model
model.fit(train_images, train_labels, epochs=10)

# Evauluating model's accuracy
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)

Epoch 1/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.5051 - accuracy: 0.8230
Epoch 2/10
1875/1875 [==============================] - 40s 22ms/step - loss: 0.3742 - accuracy: 0.8650
Epoch 3/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.3366 - accuracy: 0.8779
Epoch 4/10
1875/1875 [==============================] - 35s 19ms/step - loss: 0.3129 - accuracy: 0.8845
Epoch 5/10
1875/1875 [==============================] - 33s 18ms/step - loss: 0.2954 - accuracy: 0.8907
Epoch 6/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.2817 - accuracy: 0.8957
Epoch 7/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.2681 - accuracy: 0.9009
Epoch 8/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.2578 - accuracy: 0.9037
Epoch 9/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.2500 - accuracy: 0.9068
Epoch 10/10
1875/1875 [==============================] - 35s 19m

Simple model results
Test accuracy: 0.8819000124931335 
Train accuracy: 0.9170666933059692

## Change number of neurons in the hidden layer and compare results.

In [17]:
neurons_qty = [1, 32, 64, 256, 512]

for neurons in neurons_qty:
    
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(28, 28)),
        keras.layers.Dense(neurons, activation='relu'),
        keras.layers.Dense(10)
    ])

    model.compile(optimizer='adam',
                  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                  metrics=['accuracy'])

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
    train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
    print('\nTest and train accuracy:', test_acc, train_acc)
    print('Quantity of neurons:', neurons)
    print()



Epoch 1/10
1875/1875 [==============================] - 16s 8ms/step - loss: 1.9632 - accuracy: 0.1944
Epoch 2/10
1875/1875 [==============================] - 14s 7ms/step - loss: 1.6922 - accuracy: 0.2760
Epoch 3/10
1875/1875 [==============================] - 14s 7ms/step - loss: 1.5913 - accuracy: 0.3180
Epoch 4/10
1875/1875 [==============================] - 15s 8ms/step - loss: 1.5527 - accuracy: 0.3335
Epoch 5/10
1875/1875 [==============================] - 17s 9ms/step - loss: 1.5296 - accuracy: 0.3420
Epoch 6/10
1875/1875 [==============================] - 13s 7ms/step - loss: 1.5133 - accuracy: 0.3431
Epoch 7/10
1875/1875 [==============================] - 12s 7ms/step - loss: 1.5020 - accuracy: 0.3472
Epoch 8/10
1875/1875 [==============================] - 16s 9ms/step - loss: 1.4930 - accuracy: 0.3463
Epoch 9/10
1875/1875 [==============================] - 15s 8ms/step - loss: 1.4850 - accuracy: 0.3492
Epoch 10/10
1875/1875 [==============================] - 13s 7ms/step - l

Simple model results (128 neurons)
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Quantity of neurons: 1
Test accuracy: 0.36739999055862427 
Train accuracy: 0.3657500147819519

Quantity of neurons: 32
Test accuracy: 0.8707000017166138
Train accuracy: 0.8974999785423279

Quantity of neurons: 64
Test accuracy: 0.8787000179290771
Train accuracy: 0.90969997644424447

Quantity of neurons: 256
Test accuracy: 0.8913999795913696
Train accuracy: 0.9236833453178406

Quantity of neurons: 512
Test accuracy: 0.8873999714851379
Train accuracy: 0.9239833354949951

Best train results: 0.9239833354949951 (512 neurons)
Best test results: 0.8913999795913696 (256 neurons)

Worse train and test results: 1 neuron

Best difference between test and train results: 0.0016499757766723633 (1 neuron)
Worse difference between test and train results: 0.03658336400985718 (512 neurons)

**Conclusions**
- Increasing number of neurons gives better results on training set.
But in the same time it can cause overfitting.
- Low number of neurons is a reason of underfitting.

<b style='color:red;'>FROM CHATGPT</b>

Adjusting the number of neurons in a neural network layer can affect the network's capacity and its ability to learn complex patterns in the data. Here's how the number of neurons can influence the model:

1. **Underfitting vs. Overfitting**:
   
   - **Too Few Neurons**: If you have too few neurons in your network, it may not have enough capacity to capture the underlying patterns in the data. This can lead to underfitting, where the model cannot learn the training data well and performs poorly on both the training and validation sets.

   - **Too Many Neurons**: On the other hand, if you have too many neurons, the network may have excessive capacity and can memorize the training data, leading to overfitting. Overfit models perform very well on the training data but generalize poorly to unseen data.

2. **Training Time**:
   
   - **More Neurons**: Increasing the number of neurons typically leads to longer training times. A network with more neurons requires more computations and may require more data to train effectively.

3. **Complexity of Learned Features**:
   
   - **More Neurons**: A network with more neurons can potentially learn more complex and fine-grained features from the data, which can be beneficial for tasks with intricate patterns.

   - **Fewer Neurons**: Conversely, a network with fewer neurons may generalize better when the dataset is small or when simpler features are sufficient for the task.

## Try to add additional hidden layer and compare results

In [18]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
                loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

Epoch 1/10
1875/1875 [==============================] - 43s 22ms/step - loss: 0.4865 - accuracy: 0.8251
Epoch 2/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.3638 - accuracy: 0.8649
Epoch 3/10
1875/1875 [==============================] - 43s 23ms/step - loss: 0.3274 - accuracy: 0.8786
Epoch 4/10
1875/1875 [==============================] - 46s 25ms/step - loss: 0.3063 - accuracy: 0.8868
Epoch 5/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.2871 - accuracy: 0.8925
Epoch 6/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.2728 - accuracy: 0.8980
Epoch 7/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.2588 - accuracy: 0.9024
Epoch 8/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.2488 - accuracy: 0.9063
Epoch 9/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.2377 - accuracy: 0.9100
Epoch 10/10
1875/1875 [==============================] - 40s 21m

Simple model results (128 neurons)
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Additional hidden layer:
Test accuracy: 0.8787999749183655
Train accuracy: 0.9135666489601135

**Conclusion**
Additional hidden layer gave lower accuracy and less stable results

<b style='color:red;'>FROM CHATGPT</b>

The number of hidden layers determines the depth of the network, and it's often referred to as the network's depth or depth of the architecture. Here's how the number of hidden layers can influence the results:

1. **Model Capacity**:
   
   - **More Hidden Layers**: Increasing the number of hidden layers increases the model's capacity to learn complex patterns in the data. Deeper networks can potentially capture hierarchical and abstract features, which can be beneficial for tasks with intricate patterns.

   - **Fewer Hidden Layers**: Conversely, having fewer hidden layers may limit the model's capacity, and it might struggle to represent complex relationships in the data.

2. **Overfitting vs. Underfitting**:
   
   - **More Hidden Layers**: Deeper networks with more hidden layers are more prone to overfitting, especially when dealing with small datasets. They have a greater potential to memorize the training data, leading to poor generalization.

   - **Fewer Hidden Layers**: Simpler models with fewer hidden layers are less likely to overfit, making them more suitable for small datasets or situations where a smaller model suffices.

3. **Training Time**:
   
   - **More Hidden Layers**: Deeper networks typically require longer training times because they involve more computations. Training deep models can also be computationally intensive.

   - **Fewer Hidden Layers**: Simpler models with fewer layers tend to train faster, which can be advantageous when you have limited computational resources.

4. **Data Requirements**:
   
   - **More Hidden Layers**: Deeper networks often require more data to train effectively. Having a large dataset can help prevent overfitting when using deep architectures.

   - **Fewer Hidden Layers**: Simpler models with fewer layers may perform better when data is limited.

5. **Complexity of Learned Features**:
   
   - **More Hidden Layers**: Deeper networks have the potential to learn complex and hierarchical features, which can be beneficial for tasks like image recognition, natural language processing, and speech recognition.

   - **Fewer Hidden Layers**: Simpler models may be sufficient for tasks where the relationships in the data are relatively straightforward.

6. **Regularization**:
   
   - **More Hidden Layers**: Deeper networks may require stronger regularization techniques (e.g., dropout, L1/L2 regularization) to prevent overfitting.

   - **Fewer Hidden Layers**: Simpler models may need less regularization.



## Change activation function of hidden layer to sigmoid and compare results

In [4]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='sigmoid'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

2023-09-25 15:49:23.682492: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.


Epoch 1/10
1875/1875 [==============================] - 52s 27ms/step - loss: 0.5440 - accuracy: 0.8135
Epoch 2/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.3929 - accuracy: 0.8583
Epoch 3/10
1875/1875 [==============================] - 48s 26ms/step - loss: 0.3547 - accuracy: 0.8717
Epoch 4/10
1875/1875 [==============================] - 45s 24ms/step - loss: 0.3312 - accuracy: 0.8803
Epoch 5/10
1875/1875 [==============================] - 45s 24ms/step - loss: 0.3118 - accuracy: 0.8869
Epoch 6/10
1875/1875 [==============================] - 47s 25ms/step - loss: 0.2964 - accuracy: 0.8916
Epoch 7/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.2839 - accuracy: 0.8953
Epoch 8/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.2722 - accuracy: 0.8996
Epoch 9/10
1875/1875 [==============================] - 42s 23ms/step - loss: 0.2636 - accuracy: 0.9031
Epoch 10/10
1875/1875 [==============================] - 39s 21m

Simple model results (relu activator)
Test accuracy: 0.8819000124931335 
Train accuracy: 0.9170666933059692

Results of the network with sigmoid activator
Test accuracy: 0.8819000124931335
Train accuracy: 0.9121333360671997

**Conclusion**
Relu and sigmoid activators showed almost the same results

<b style='color:red;'>FROM CHATGPT</b>

1. **ReLU (Rectified Linear Unit)**:

   - **Advantages**:
     - **Non-linearity**: ReLU introduces non-linearity into the model, allowing the network to learn complex relationships in the data.
     - **Sparsity**: ReLU activations are sparse because they output zero for negative inputs, which can make the network more efficient and easier to train.
     - **Gradient Descent**: The derivative of ReLU is either 0 or 1, making it easier for gradient-based optimization algorithms to converge compared to saturating activation functions like sigmoid.

   - **Use Cases**:
     - ReLU is commonly used as an activation function in hidden layers of deep neural networks, especially in deep convolutional neural networks (CNNs) used for image processing tasks.
     - It is well-suited for problems with large amounts of data and complex patterns.

   - **Drawbacks**:
     - **Dead Neurons**: ReLU units can sometimes become "dead" (outputting zero for all inputs) during training if they encounter large gradients. This issue can be addressed with variants like Leaky ReLU and Parametric ReLU (PReLU).

2. **Sigmoid**:

   - **Advantages**:
     - **Smoothness**: Sigmoid is a smooth, differentiable activation function that produces values between 0 and 1. It can be useful for ensuring smooth transitions in the network's outputs.
     - **Use in Output Layer**: Sigmoid is often used in the output layer of binary classification models because it squashes the network's output to the range [0, 1], representing probabilities.

   - **Use Cases**:
     - Sigmoid is historically used in shallow neural networks and logistic regression models.
     - It is suitable for binary classification problems when the output needs to be interpreted as probabilities.

   - **Drawbacks**:
     - **Vanishing Gradients**: Sigmoid tends to suffer from the vanishing gradient problem, especially in deep networks, where gradients can become very small during backpropagation, making training slow and difficult.
     - **Saturated Neurons**: Sigmoid activations saturate when the inputs are too far from zero, causing the gradients to become close to zero. This can lead to slow convergence.

In summary, the choice between ReLU and sigmoid activations depends on the specific problem you are solving and the architecture of your neural network:

- ReLU is the default choice for hidden layers in most modern neural networks because of its non-linearity, efficiency, and ability to mitigate vanishing gradient problems. Variants like Leaky ReLU and PReLU can be used to address the "dead neuron" issue.

- Sigmoid is still useful in specific scenarios, such as binary classification problems and shallow networks. However, it is less commonly used in deep learning due to its drawbacks, especially in deep architectures.

## Use RMSprop optimizer instead of Adam and compare results

In [20]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='RMSprop',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

Epoch 1/10
1875/1875 [==============================] - 35s 18ms/step - loss: 0.5028 - accuracy: 0.8204
Epoch 2/10
1875/1875 [==============================] - 33s 18ms/step - loss: 0.3740 - accuracy: 0.8652
Epoch 3/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.3414 - accuracy: 0.8782
Epoch 4/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.3214 - accuracy: 0.8850
Epoch 5/10
1875/1875 [==============================] - 30s 16ms/step - loss: 0.3091 - accuracy: 0.8900
Epoch 6/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.2985 - accuracy: 0.8956
Epoch 7/10
1875/1875 [==============================] - 34s 18ms/step - loss: 0.2912 - accuracy: 0.8993
Epoch 8/10
1875/1875 [==============================] - 33s 18ms/step - loss: 0.2813 - accuracy: 0.9015
Epoch 9/10
1875/1875 [==============================] - 33s 17ms/step - loss: 0.2756 - accuracy: 0.9039
Epoch 10/10
1875/1875 [==============================] - 32s 17m

Simple model results (adam optimizer)
Test accuracy: 0.8819000124931335 
Train accuracy: 0.9170666933059692

Results of the network with RMSprop optimizer
Test accuracy: 0.8837000131607056
Train accuracy: 0.915149986743927
 
**Conclusion**
Adam and RMSprop optimizers showed almost the same results

<b style='color:red;'>FROM CHATGPT</b>

Adam (short for Adaptive Moment Estimation) and RMSprop (short for Root Mean Square Propagation) are both popular optimization algorithms used in training neural networks. They share some similarities but also have distinct differences in their behavior and impact on training. Let's explore the key differences and impacts of these two optimizers:

**Adam Optimizer**:

1. **Adaptive Learning Rates**:
   
   - Adam computes adaptive learning rates for each parameter. It adjusts the learning rates based on the past gradients of each parameter, allowing for faster convergence in some cases.

2. **Momentum-like Update**:
   
   - Adam includes a momentum-like term that helps accelerate convergence, especially in the presence of noisy gradients. It combines the concepts of momentum (like SGD with momentum) and RMSprop.

3. **Bias Correction**:
   
   - Adam includes bias correction terms to account for the fact that the moving averages of past gradients are initialized with zeros. This bias correction helps improve the accuracy of parameter updates, especially in the early stages of training.

4. **Hyperparameters**:
   
   - Adam has hyperparameters such as the learning rate (`lr`), β1 (momentum-like term decay rate), and β2 (decay rate for the moving average of squared gradients). These hyperparameters need to be tuned for optimal performance.

**RMSprop Optimizer**:

1. **Root Mean Square Gradients**:
   
   - RMSprop computes a moving average of the squared gradients for each parameter. It uses this moving average to scale the learning rates for each parameter individually.

2. **No Momentum Term**:
   
   - Unlike Adam, RMSprop does not have an explicit momentum term. It relies solely on adaptive learning rates to adjust the step sizes for parameter updates.

3. **Hyperparameters**:
   
   - RMSprop has hyperparameters such as the learning rate (`lr`) and a decay factor (usually denoted as γ or β) for the moving average of squared gradients. These hyperparameters also require tuning.

**Differences and Impact**:

1. **Convergence Speed**:
   
   - Adam often converges faster than RMSprop due to its adaptive learning rates and momentum-like term. It can handle noisy or sparse gradients more effectively.

2. **Stability**:
   
   - RMSprop can be more stable and robust in some cases because it doesn't include the momentum-like term, which can sometimes lead to oscillations in parameter updates.

3. **Hyperparameter Sensitivity**:
   
   - Adam may be more sensitive to the choice of hyperparameters, and poorly chosen hyperparameters can lead to erratic behavior during training.

4. **Common Use Cases**:
   
   - Adam is commonly used as a default optimizer for many deep learning applications because of its overall good performance and convergence properties.

5. **Experimentation**:
   
   - The choice between Adam and RMSprop may require experimentation. In some cases, one optimizer may outperform the other depending on the specific problem and architecture.

In summary, both Adam and RMSprop are effective optimizers, and the choice between them depends on your specific problem, dataset, and experimentation. Adam often converges faster, while RMSprop can be more stable. Hyperparameter tuning is crucial for optimizing the performance of either optimizer.


## Change number of epochs to 5 and compare results

In [21]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

Epoch 1/5
1875/1875 [==============================] - 40s 21ms/step - loss: 0.4982 - accuracy: 0.8261
Epoch 2/5
1875/1875 [==============================] - 35s 19ms/step - loss: 0.3761 - accuracy: 0.8645
Epoch 3/5
1875/1875 [==============================] - 35s 19ms/step - loss: 0.3351 - accuracy: 0.8758
Epoch 4/5
1875/1875 [==============================] - 32s 17ms/step - loss: 0.3108 - accuracy: 0.8864
Epoch 5/5
1875/1875 [==============================] - 35s 19ms/step - loss: 0.2923 - accuracy: 0.8921
313/313 - 2s - loss: 0.3684 - accuracy: 0.8697 - 2s/epoch - 5ms/step
1875/1875 - 9s - loss: 0.2851 - accuracy: 0.8951 - 9s/epoch - 5ms/step

Test and train accuracy: 0.869700014591217 0.8951333165168762


Simple model results (10 epochs)
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Results of the network with 5 epochs
Test accuracy: 0.869700014591217 
Train accuracy: 0.8951333165168762

**Conclusion**
Accorucy got lower after number of epoches was redused from 10 to 5 => *underfitting*

<b style='color:red;'>FROM CHATGPT</b>

1. **Underfitting vs. Overfitting**:

   - **Fewer Epochs**:
     - If you use too few epochs, the model may not have enough training cycles to learn the underlying patterns in the data. This can result in underfitting, where the model's performance is suboptimal both on the training and validation datasets.
   
   - **More Epochs**:
     - Using more epochs allows the model to continue learning from the data. It can help reduce underfitting and improve the model's performance on the training dataset.

   - **Risk of Overfitting**:
     - However, increasing the number of epochs beyond a certain point can lead to overfitting. Overfitting occurs when the model starts memorizing the training data instead of generalizing from it. This can lead to poor performance on unseen data (validation or test sets).

2. **Validation Loss and Accuracy**:

   - **Early Stopping**:
     - Monitoring the validation loss and accuracy during training is essential. If you notice that the validation loss starts to increase or the validation accuracy plateaus or decreases while the training loss continues to decrease, it may indicate that you've trained for too many epochs, and early stopping might be beneficial.

3. **Computational Resources**:

   - **Training Time**:
     - Using a large number of epochs can increase the training time, especially for deep and complex models. It's important to strike a balance between the number of epochs and available computational resources.

4. **Dataset Size**:

   - **Impact on Small Datasets**:
     - The impact of the number of epochs is more pronounced on smaller datasets. Small datasets tend to overfit more easily, so using an appropriate number of epochs is crucial.

5. **Learning Rate Scheduling**:

   - **Learning Rate Adaptation**:
     - Some training scenarios involve using learning rate scheduling or dynamic learning rates. In these cases, the number of epochs can be related to how the learning rate is adjusted during training.

6. **Hyperparameter Tuning**:

   - **Hyperparameter Interaction**:
     - The number of epochs can interact with other hyperparameters, such as the learning rate, batch size, and regularization strength. Fine-tuning these hyperparameters may require adjusting the number of epochs.

## Use dropout to avoid overfitting and compare results

In [22]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

Epoch 1/10
1875/1875 [==============================] - 33s 17ms/step - loss: 0.6278 - accuracy: 0.7770
Epoch 2/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.4773 - accuracy: 0.8287
Epoch 3/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.4445 - accuracy: 0.8394
Epoch 4/10
1875/1875 [==============================] - 35s 19ms/step - loss: 0.4279 - accuracy: 0.8448
Epoch 5/10
1875/1875 [==============================] - 35s 19ms/step - loss: 0.4118 - accuracy: 0.8508
Epoch 6/10
1875/1875 [==============================] - 35s 18ms/step - loss: 0.4017 - accuracy: 0.8539
Epoch 7/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.3935 - accuracy: 0.8559
Epoch 8/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.3866 - accuracy: 0.8575
Epoch 9/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.3777 - accuracy: 0.8607
Epoch 10/10
1875/1875 [==============================] - 36s 19m

Simple model results 
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Results of the network using dropout
Test accuracy: 0.8719000220298767
Train accuracy: 0.8901333212852478

**Conclusion**
Results are slightly worse but more stable

<b style='color:red;'>FROM CHATGPT</b>
Dropout is a regularization technique commonly used in neural networks to prevent overfitting. It has a significant impact on the training and performance of a model. Dropout works by randomly deactivating (dropping out) a fraction of neurons during each training batch, forcing the model to learn more robust and generalizable features. Here's how dropout can impact your neural network:

1. **Reduction in Overfitting**:

   - **Impact**: Dropout is primarily used to reduce overfitting. It prevents the network from relying too heavily on specific neurons and encourages it to learn a more diverse set of features.

   - **Result**: By reducing overfitting, dropout can lead to improved generalization performance. The model is less likely to memorize the training data and is more likely to perform well on unseen data, such as a validation or test dataset.

2. **Regularization**:

   - **Impact**: Dropout serves as a form of regularization by adding noise to the network during training. This regularization can help control the complexity of the model and prevent it from fitting the training data too closely.

   - **Result**: A model with dropout is less likely to suffer from high variance, which can lead to better convergence during training.

3. **Ensemble Effect**:

   - **Impact**: Dropout can be seen as training an ensemble of multiple neural networks with shared weights. During each training batch, different neurons are dropped out, effectively training different subnetworks.

   - **Result**: The ensemble effect can improve the model's robustness and ability to generalize. It's as if the network has learned to make predictions by considering the opinions of multiple experts (the different subnetworks).

4. **Training Time**:

   - **Impact**: Dropout can increase training time because it requires multiple forward and backward passes through the network during each epoch. This is because dropout is active during training but not during inference (making predictions).

   - **Result**: Longer training time is a trade-off for the benefits of dropout, but it is often a worthwhile trade-off to achieve better generalization.

5. **Tuning Dropout Rate**:

   - **Impact**: The choice of dropout rate (the fraction of neurons to drop) is a hyperparameter that impacts the model's behavior.

   - **Result**: The optimal dropout rate may vary depending on the dataset and network architecture. Too high of a dropout rate can lead to underfitting, while too low of a dropout rate may not provide sufficient regularization.

6. **Use in Specific Layers**:

   - **Impact**: Dropout is typically applied to hidden layers rather than input or output layers.

   - **Result**: This allows the model to maintain more information at the input and output layers while adding regularization and reducing overfitting in the intermediate layers.


## Try different normalizations such as BatchNormalization and compare results

In [23]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
print('\nTest and train accuracy:', test_acc, train_acc)
print()

Epoch 1/10
1875/1875 [==============================] - 42s 22ms/step - loss: 0.4912 - accuracy: 0.8279
Epoch 2/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.4052 - accuracy: 0.8560
Epoch 3/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.3757 - accuracy: 0.8659
Epoch 4/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.3484 - accuracy: 0.8732
Epoch 5/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.3359 - accuracy: 0.8781
Epoch 6/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.3216 - accuracy: 0.8838
Epoch 7/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.3108 - accuracy: 0.8862
Epoch 8/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.3045 - accuracy: 0.8887
Epoch 9/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.2945 - accuracy: 0.8910
Epoch 10/10
1875/1875 [==============================] - 37s 20m

Simple model results
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Results of the network using dropout
Test accuracy: 0.8737000226974487
Train accuracy: 0.9031500220298767
 
**Conclusion**
Results are worse

<b style='color:red;'>FROM CHATGPT</b>
Batch Normalization (BatchNorm) is a layer commonly used in neural networks, and it has a significant impact on the training and performance of a model. Here's how Batch Normalization can impact your neural network:

1. **Improved Training Speed**:

   - **Impact**: BatchNorm can accelerate training by reducing internal covariate shift. It normalizes the activations of a layer, ensuring that they have a consistent mean and variance across mini-batches during training.

   - **Result**: Faster convergence is often observed, allowing you to train deeper networks or achieve the same level of performance with fewer training epochs.

2. **Reduced Sensitivity to Initialization**:

   - **Impact**: BatchNorm makes networks less sensitive to weight initialization. This means you can use higher learning rates without causing instability during training.

   - **Result**: Faster training convergence and more robust training dynamics.

3. **Mitigation of Vanishing/Exploding Gradient Problems**:

   - **Impact**: BatchNorm can help mitigate the vanishing gradient problem by normalizing activations. This is particularly beneficial in deep networks.

   - **Result**: Deeper networks can be trained more effectively, leading to better model performance.

4. **Regularization Effect**:

   - **Impact**: BatchNorm acts as a form of regularization by adding noise to activations. This can help reduce overfitting.

   - **Result**: Models with BatchNorm layers tend to be more robust to overfitting, allowing for the training of larger and more complex networks.

5. **Smoothing of Loss Landscape**:

   - **Impact**: BatchNorm can have a smoothing effect on the loss landscape during training, making optimization more stable.

   - **Result**: This helps the model to avoid getting stuck in poor local minima during optimization.

6. **Impact on Learning Rate Scheduling**:

   - **Impact**: BatchNorm can affect the choice of learning rate scheduling. With BatchNorm, it's often possible to use higher initial learning rates and more aggressive learning rate schedules.

   - **Result**: This can lead to faster convergence and better generalization.

7. **Use in Convolutional and Fully Connected Layers**:

   - **Impact**: BatchNorm can be used in both convolutional and fully connected layers, making it versatile and applicable to a wide range of neural network architectures.

   - **Result**: It can improve the training dynamics and generalization of various types of networks, including convolutional neural networks (CNNs) and feedforward neural networks.

8. **Impact on Mini-Batch Size**:

   - **Impact**: BatchNorm's effectiveness can depend on the mini-batch size used during training. It typically works best with reasonably sized mini-batches.

   - **Result**: The choice of mini-batch size should be considered when using BatchNorm.

## Try different types of weight initializing (i.e., he_normal or glorot_uniform) and compare results.

In [24]:
relu_initializers = ['he_normal', 'random_normal', 'random_uniform', 'truncated_normal']

In [25]:
for initializer in relu_initializers:

    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(28, 28)),
        keras.layers.Dense(128, activation='relu', kernel_initializer=initializer),
        keras.layers.Dense(10)
    ])

    model.compile(optimizer='adam',
                  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                  metrics=['accuracy'])

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
    train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
    print('\nTest and train accuracy:', test_acc, train_acc)
    print('Initializer:', initializer)
    print()

Epoch 1/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.4996 - accuracy: 0.8226
Epoch 2/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.3772 - accuracy: 0.8653
Epoch 3/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.3389 - accuracy: 0.8773
Epoch 4/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.3136 - accuracy: 0.8847
Epoch 5/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.2961 - accuracy: 0.8899
Epoch 6/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.2789 - accuracy: 0.8962
Epoch 7/10
1875/1875 [==============================] - 44s 24ms/step - loss: 0.2671 - accuracy: 0.9013
Epoch 8/10
1875/1875 [==============================] - 42s 23ms/step - loss: 0.2582 - accuracy: 0.9042
Epoch 9/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.2483 - accuracy: 0.9072
Epoch 10/10
1875/1875 [==============================] - 40s 21m

Simple model results
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Results of the network using 'he_normal' weight initializer
Test accuracy: 0.879800021648407
Train accuracy: 0.9140666723251343

Results of the network using 'random_normal' weight initializer
Test accuracy: 0.878000020980835
Train accuracy: 0.9124333262443542

Results of the network using 'random_uniform' weight initializer
Test accuracy: 0.8781999945640564
Train accuracy: 0.9146999716758728

Results of the network using 'truncated_normal' weight initializer
Test accuracy: 0.8805999755859375
Train accuracy: 0.9150833487510681

**Conclusion**
No obvious difference between results of 'he_normal', 'random_normal', 'random_uniform', 'truncated_normal' weight initializers.

<b style='color:red;'>FROM CHATGPT</b>
The choice of weight initialization method in a neural network can have a significant impact on the training and performance of the model. Each weight initialization method initializes the neural network weights differently, influencing the convergence speed and final results. Here's an overview of the impact of some commonly used weight initializers:

1. **'he_normal' (He Normal Initialization)**:

   - **Impact**: He Normal Initialization initializes weights with a mean of 0 and a standard deviation calculated based on the number of input units in the layer. It is specifically designed for ReLU activation functions.

   - **Result**: He Normal Initialization can accelerate convergence and help prevent vanishing gradients when using ReLU activations. It is a popular choice for deep networks and is known for its effectiveness.

2. **'random_normal' (Random Normal Initialization)**:

   - **Impact**: Random Normal Initialization initializes weights with random values sampled from a normal distribution with a specified mean and standard deviation.

   - **Result**: This initialization provides a starting point for training and allows the network to learn weight values during training. However, the effectiveness can depend on the choice of mean and standard deviation.

3. **'random_uniform' (Random Uniform Initialization)**:

   - **Impact**: Random Uniform Initialization initializes weights with random values sampled from a uniform distribution within a specified range.

   - **Result**: Like 'random_normal,' this initialization provides an initial range of weights for training. It can be useful, but it may require careful tuning of the range to ensure effective training.

4. **'truncated_normal' (Truncated Normal Initialization)**:

   - **Impact**: Truncated Normal Initialization is similar to 'random_normal' but ensures that weights are sampled from a normal distribution with values within a specified range, effectively truncating any weights outside that range.

   - **Result**: It can be useful for preventing very large or very small weight values, which can be beneficial for network stability.

In summary:

- **'he_normal'** is often a good choice when using ReLU activations, particularly in deep networks. It helps mitigate issues like vanishing gradients and can lead to faster convergence.

- **'random_normal,' 'random_uniform,'** and **'truncated_normal'** provide randomness in weight initialization, allowing the network to adapt and learn weight values during training. However, their effectiveness may require careful tuning of parameters such as the mean, standard deviation, or range.

## Use L1 and L2 regulizations and compare results

In [27]:
regularizers = [l1(0.01), l1(0.1), l1(1), l2(0.01), l2(0.1), l2(1)]

In [28]:
for regulizer in regularizers:

    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(28, 28)),
        keras.layers.Dense(128, activation='relu', kernel_regularizer=regulizer),
        keras.layers.Dense(10)
    ])

    model.compile(optimizer='adam',
                  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                  metrics=['accuracy'])

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
    train_loss, train_acc = model.evaluate(train_images,  train_labels, verbose=2)
    print('\nTest and train accuracy:', test_acc, train_acc)
    print('Regulizer:', regulizer)
    print()

Epoch 1/10
1875/1875 [==============================] - 42s 22ms/step - loss: 2.2552 - accuracy: 0.7111
Epoch 2/10
1875/1875 [==============================] - 41s 22ms/step - loss: 1.2192 - accuracy: 0.7553
Epoch 3/10
1875/1875 [==============================] - 39s 21ms/step - loss: 1.1523 - accuracy: 0.7653
Epoch 4/10
1875/1875 [==============================] - 43s 23ms/step - loss: 1.1220 - accuracy: 0.7691
Epoch 5/10
1875/1875 [==============================] - 42s 22ms/step - loss: 1.0958 - accuracy: 0.7723
Epoch 6/10
1875/1875 [==============================] - 38s 20ms/step - loss: 1.0853 - accuracy: 0.7754
Epoch 7/10
1875/1875 [==============================] - 39s 21ms/step - loss: 1.0714 - accuracy: 0.7770
Epoch 8/10
1875/1875 [==============================] - 38s 20ms/step - loss: 1.0599 - accuracy: 0.7786
Epoch 9/10
1875/1875 [==============================] - 38s 20ms/step - loss: 1.0579 - accuracy: 0.7796
Epoch 10/10
1875/1875 [==============================] - 37s 20m

Simple model results
Test accuracy: 0.8819000124931335
Train accuracy: 0.9170666933059692

Network results with l1(0.01)
Test accuracy: 0.7720000147819519
Train accuracy: 0.7795666456222534

Network results with l1(0.1)
Test accuracy: 0.6478000283241272
Train accuracy: 0.6492000222206116

Network results with l1(1)
Test accuracy: 0.17470000684261322
Train accuracy: 0.17581667006015778

Network results with l2(0.01)
Test accuracy: 0.8295000195503235 
Train accuracy: 0.8430333137512207

Network results with l2(0.1)
Test accuracy: 0.7785000205039978
Train accuracy: 0.7924833297729492
 
Network results with l2(1)
Test accuracy: 0.6452000141143799
Train accuracy: 0.6490833163261414

**Conclusion**
L1 and L2 regulizations definitely makes a network more stable.
But results both on train and test sets got worse.

<b style='color:red;'>FROM CHATGPT</b>
L1 and L2 regularization are techniques used to prevent overfitting and improve the generalization performance of neural networks by adding penalty terms to the loss function. These penalty terms encourage the model to have smaller weights. Here's an overview of the impact of L1 and L2 regularization and the impact of their values:

1. **L1 Regularization (Lasso Regularization)**:

   - **Impact**: L1 regularization adds a penalty term to the loss function that is proportional to the absolute values of the weights. It encourages sparsity in the weight vectors, effectively driving some weights to zero.

   - **Result**: The impact of L1 regularization is that it can lead to feature selection, where some weights are set to exactly zero, effectively removing certain features from consideration. This can simplify the model and make it more interpretable.

   - **L1 Strength (Lambda)**: The strength of L1 regularization is controlled by a hyperparameter lambda (λ). Larger values of λ result in more aggressive weight shrinking and sparsity. Tuning λ is crucial, as it determines the trade-off between model complexity and regularization.

2. **L2 Regularization (Ridge Regularization)**:

   - **Impact**: L2 regularization adds a penalty term to the loss function that is proportional to the square of the weights. It discourages large weights but doesn't drive weights to exactly zero.

   - **Result**: The impact of L2 regularization is that it encourages the model to distribute the weight values more evenly across all features, preventing the dominance of a few features. This often results in smoother loss landscapes and can be more forgiving to individual outliers.

   - **L2 Strength (Lambda)**: Similar to L1, the strength of L2 regularization is controlled by a hyperparameter λ. Larger values of λ lead to stronger regularization.

Key takeaways regarding the impact of L1 and L2 regularization:

- **Overfitting Prevention**: Both L1 and L2 regularization help prevent overfitting by controlling the complexity of the model.

- **Feature Selection**: L1 regularization can lead to feature selection by driving some weights to zero, making it useful for feature selection and model interpretability.

- **L2 Norm vs. L1 Norm**: L2 regularization encourages all weight values to be small, while L1 regularization encourages some weights to be exactly zero.

- **Hyperparameter Tuning**: The choice of the regularization strength (λ) is critical and depends on the specific problem and dataset. It is typically determined through cross-validation or other hyperparameter tuning techniques.